core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -710,7 +723,11 @@ object SparkSubmit {
       printWarning("Subclasses of scala.App may not work correctly. Use a main() method instead.")
     }
 
-    val mainMethod = mainClass.getMethod("main", new Array[String](0).getClass)
+    val mainMethod = if (threadEnabled ) {
+      mainClass.getMethods().filter(i => i.getName() == "sparkMain")(0)
```

Comment:
nit: either

    `.filter { i => ... }`

or

    `.filter(_.getName() == ...)`

Also, theoretically you shouldn't need to use reflection here since you have a proper type. Scala singletons make this a little more interesting, since I'm not sure how to get the singleton using reflection, but anyway, this can be changed later.